pcdsdevices/targets.py

```diff
@@ -260,3 +267,795 @@ def move(self, nxspaces, nyspaces, wait=False):
+
+        return m_points, n_points, coeffs
+
+    def save_grid(self, sample_name, path=None):
+        """
+        Save a grid of mapped points for a sample.
+
+        This will save the date it was created, along with the sample name,
+        the m and n points, the coordinates for the four corners, and the
+        coefficients that will help get the x and y position on the grid.
+        If an existing name for a sample is saved again, it will override
+        the information for that sample.
+
+        Parameters
+        ----------
+        sample_name : int
+            A name to identify the sample grid, should be snake_case style.
+        path : str, optional
+            Path to the `.yml` file. Defaults to the path defined when
+            creating this object.
+
+        Examples
+        --------
+        >>> save_grid('sample_1')
+        """
+        path = path or self._path
+        now = str(datetime.now())
+        top_left, top_right, bottom_right, bottom_left = (), (), (), ()
+        if self.get_presets():
+            top_left, top_right, bottom_right, bottom_left = self.get_presets()
+        m_points, n_points = self.m_n_points
+        coefficients = self.coefficients
+        data = {sample_name: {"time_created": now,
+                              "top_left": top_left,
+                              "top_right": top_right,
+                              "bottom_right": bottom_right,
+                              "bottom_left": bottom_left,
+                              "M": m_points,  # number of rows
+                              "N": n_points,  # number of columns
+                              "coefficients": coefficients}}
+        try:
+            temp_data = json.loads(json.dumps(data[sample_name]))
+            jsonschema.validate(temp_data, self.sample_schema)
+        except jsonschema.exceptions.ValidationError as err:
+            logger.warning('Invalid input: %s', err)
+        # override the existing sample grids or append other grids
+        # TODO: how do i prevent from writing something bad and wiping the
+        # whole file off - check for None values???
+        with open(path) as sample_file:
+            yaml_dict = yaml.safe_load(sample_file) or {}
+            yaml_dict.update(data)
+        with open(path, 'w') as sample_file:
+            yaml.safe_dump(yaml_dict, sample_file,
+                           sort_keys=False, default_flow_style=False)
+
+    # TODO: maybe I shouold not have this here...
+    def remove_all_samples(self, path=None):
+        """
+        Empty the samples file.
+
+        All the grid samples will be deleted.
+
+        Parameters
+        ----------
+        path : string, optional
+            Path of .yml file with samples.
+        """
+        path = path or self._path
+        with open(path, 'w') as sample_file:
+            yaml.safe_dump(None, sample_file)
+
+    def mesh_interpolation(self, top_left, top_right, bottom_right,
```

Comment:
I have a few functions here that probably can be moved to the `pcdscalc` - 
- `mesh_interpolation`
- `get_unit_meshgrid`
- `convert_to_physical`
- maybe others....

They are however pretty specific to this `XYGridStage` object, should I still move them?

Reply:
I think it's correct to keep things that are very specific to this class here- after all, if it's only this class that wants to use them, then we have to track down a central library if they need an update.

When we find ourselves wanting to re-use the code for something else, that's when it's time to send it to `pcdscalc`.

Reply:
Ok that sounds good. If I go about calculating each point  during a scan - then some of these functions might be useful in the `nabs` as well - I am currently trying this in my scan function in `nabs`